native_client/deepspeech.cc

```diff
@@ -8,45 +8,423 @@
 using tensorflow::ctc::CTCBeamSearchDecoder;
 using tensorflow::ctc::CTCDecoder;
 
 namespace DeepSpeech {
 
+class StreamingState {
+public:
+  std::vector<float> accumulated_logits;
+  short audio_buffer[AUDIO_WIN_LEN_SAMPLES];
+  unsigned int audio_buffer_len;
+  float mfcc_buffer[MFCC_FEATS_PER_TIMESTEP];
+  unsigned int mfcc_buffer_len;
+  float batch_buffer[N_STEPS_PER_BATCH*MFCC_FEATS_PER_TIMESTEP];
+  unsigned int batch_buffer_len;
+  bool skip_next_mfcc;
+};
+
 class Private {
-  public:
-    MemmappedEnv* mmap_env;
-    Session* session;
-    GraphDef graph_def;
-    int ncep;
-    int ncontext;
-    Alphabet* alphabet;
-    KenLMBeamScorer* scorer;
-    int beam_width;
-    bool run_aot;
+public:
+  MemmappedEnv* mmap_env;
+  Session* session;
+  GraphDef graph_def;
+  int ncep;
+  int ncontext;
+  Alphabet* alphabet;
+  KenLMBeamScorer* scorer;
+  int beam_width;
+  bool run_aot;
+
+  /* This is the actual implementation of the streaming inference API, with the
+     Model class just forwarding the calls to this class.
+
+     The streaming process uses three buffers that are fed eagerly as audio data
+     is fed in. The buffers only hold the minimum amount of data needed to do a
+     step in the acoustic model. The three buffers which live in StreamingContext
+     are:
+
+     - audio_buffer, used to buffer audio samples until there's enough data to
+       compute input features for a single window.
+
+     - mfcc_buffer, used to buffer input features until there's enough data for
+       a single timestep. Remember there's overlap in the features, each timestep
+       contains N_CONTEXT past feature frames, the current feature frame, and
+       N_CONTEXT future feature frames, for a total of MFCC_WIN_LEN feature frames
+       per timestep.
+
+     - batch_buffer, used to buffer timesteps until there's enough data to compute
+       a batch of N_STEPS_PER_BATCH.
+
+     Data flows through all three buffers as audio samples are fed via the public
+     API. When audio_buffer is full, features are computed from it and pushed to
+     mfcc_buffer. When mfcc_buffer is full, the timestep is copied to batch_buffer.
+     When batch_buffer is full, we do a single step through the acoustic model
+     and accumulate results in StreamingState::accumulated_logits.
+
+     When fininshStream() is called, we decode the accumulated logits and return
+     the corresponding transcription.
+  */
+  StreamingState* setupStream(unsigned int prealloc_frames, unsigned int sample_rate);
+  void feedAudioContent(StreamingState* ctx, const short* buffer, unsigned int buffer_size);
+  char* finishStream(StreamingState* ctx);
+
+private:
+  /**
+   * @brief Perform decoding of the logits, using basic CTC decoder or
+   *        CTC decoder with KenLM enabled
+   *
+   * @param n_frames       Number of timesteps to deal with
+   * @param logits         Matrix of logits, of dimensions:
+   *                       [n_frames][batch_size][num_classes]
```

Comment:
The comment states it is a matrix of `[n_frames][batch_size][num_classes]` dimensions, but the prototype is `std::vector<float>& logits` which seems like a simple vector. Am I missing something?
Also, there is no `n_frames` as well :'(

Reply:
Good catch, I changed this from bare pointers to std::vector and forgot to update the comments.